pool.go

```diff
@@ -0,0 +1,166 @@
+package viamrtsp
+
+import (
+	"sync"
+	"time"
+
+	"go.viam.com/rdk/logging"
+)
+
+type poolItem struct {
+	frameWrapper *avFrameWrapper
+	lastAccess   time.Time
+}
+
+type framePool struct {
+	items          []poolItem
+	maxAge         time.Duration
+	preseededCount int
+
+	putCount   int
+	getCount   int
+	cleanCount int
+	newCount   int
+
+	mu           sync.RWMutex
+	stopCleaning chan struct{}
+	logger       logging.Logger
+}
+
+func newFramePool(maxAge, cleanupInterval time.Duration, preseededCount int, logger logging.Logger) *framePool {
+	pool := &framePool{
+		items:          make([]poolItem, 0, preseededCount),
+		maxAge:         maxAge,
+		preseededCount: preseededCount,
+		stopCleaning:   make(chan struct{}),
+		logger:         logger,
+	}
+
+	// Pre-seed the pool
+	for i := 0; i < preseededCount; i++ {
+		pool.items = append(pool.items, poolItem{
+			frameWrapper: pool.new(),
+			lastAccess:   time.Now(),
+		})
+	}
+
+	go pool.cleanupRoutine(cleanupInterval)
+
+	return pool
+}
+
+func (p *framePool) new() *avFrameWrapper {
+	p.logger.Debug("newFunc for pool was called!")
+	avFrame, err := newAVFrameWrapper()
+	if err != nil {
+		p.logger.Errorf("Failed to allocate AVFrame in frame pool: %v", err)
+		return nil
+	}
+	p.newCount++
+	return avFrame
+}
+
+func (p *framePool) get() *avFrameWrapper {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	if len(p.items) == 0 {
+		return nil
+	}
+
+	item := p.items[0]
+	p.items = p.items[1:]
+
+	p.logger.Debugf("Item was gotten from the pool. Len now: %d", len(p.items))
+	p.getCount++
+
+	item.frameWrapper.isInPool.Store(false)
+	return item.frameWrapper
+}
+
+func (p *framePool) put(frame *avFrameWrapper) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.items = append(p.items, poolItem{frameWrapper: frame, lastAccess: time.Now()})
+	p.logger.Debugf("Item was put in pool. Len now: %d", len(p.items))
+	p.putCount++
+
+	frame.isInPool.Store(true)
+}
+
+func (p *framePool) safelyPut(frame *avFrameWrapper) {
+	p.logger.Debug("Trying to safely put frame back into pool.")
+
+	isFreed := frame.isFreed.Load()
+	isBeingServed := frame.isBeingServed.Load()
+	isInPool := frame.isInPool.Load()
+
+	if !isFreed && !isBeingServed && !isInPool {
+		p.put(frame)
+	} else {
+		p.logger.Debugf("Frame was already freed (%t) or is currently being served (%t) or is already in the pool (%t). Cannot put.",
+			isFreed, isBeingServed, isInPool,
+		)
+	}
+}
+
+func (p *framePool) cleanupRoutine(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			p.cleanup()
+		case <-p.stopCleaning:
+			return
+		}
+	}
+}
+
+func (p *framePool) cleanup() {
```

Comment:
Now that I've gotten to the bottom here, I see the purpose/justification for a lot of these other variables. Can we just delete this extra goroutine `cleanup` for now? And replace it with a simpler algorithm? Defined in the following comment.

I don't think the complexity is solving a real problem, but it is making the behavior/resource consumption non-deterministic.

For example: suppose we have `len(p.items)` = 11 and `p.preseededCount` is 10. Based on when the cleanupRoutine ticks, we could end up with (again, for example) 1 item leftover. Or 10.

And now if we never need to create a new pool item, the long-term size of `p.items` will never converge. So if there's no guarantee we'll ever shrink to only items with a `lastAccess` within the `p.maxAge`, why bother?

Reply:
The simpler algorithm being:
```
func [safely]Put(x item) {
  if len(p.items) >= p.preseededCount {
    x.frameWrapper.free()
  }

  p.items = append(p.items, x)
}
```

Reply:
That's really nice I like that

Reply:
Added